File: 1472-design-browser-history/1472-design-browser-history.cpp
class BrowserHistory {
private:
    list<string> historyPages;
    list<string> :: iterator it;
public:
    BrowserHistory(string homepage) {
        historyPages.emplace_back(homepage);
        it = begin(historyPages);
    }
    
    void visit(string url) {
        list<string> :: iterator deleted = it;
        ++deleted;
        historyPages.erase(deleted, end(historyPages));
        historyPages.emplace_back(url);
        ++it;
        
    }
    
    string back(int steps) {
        while(it != begin(historyPages) and steps--)
            --it;
        return *it;
    }
    
    string forward(int steps) {
        while(it != (--end(historyPages)) and steps--)
            ++it;
        return *it;
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */

File: 1472-design-browser-history/1472-design-browser-history.java
class BrowserHistory {
    class Node {
        public String val;
        public Node prev;
        public Node next;
        Node(String val) {
            this.val = val;
            this.prev = null;
            this.next = null;
        }
    }
    private Node head;
    private Node current;

    public BrowserHistory(String homepage) {
        head = new Node(homepage);
        current = head;
    }
    
    public void visit(String url) {
        Node nextNode = new Node(url);
        current.next = nextNode;
        nextNode.prev = current;
        current = current.next;
    }
    
    public String back(int steps) {
        while(current != head && (steps-- > 0))
            current = current.prev;
        return current.val;
    }
    
    public String forward(int steps) {
        while(current.next != null && (steps-- > 0))
            current = current.next;
        return current.val;
    }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.visit(url);
 * String param_2 = obj.back(steps);
 * String param_3 = obj.forward(steps);
 */